### PR TITLE
Adding HP-MR Based TRISO Failure Model

### LINKED ISSUE #327: HP-MR Based TRISO Failure Model
## Model Description

Statistical TRISO failure calculation using BISON based on the irradiation condition calculated by a multiphysics HP-MR unit cell model.

It can be run a a workstation but HPC is preferred. 

## Project description

NEAMS Microreactor Application Driver Team. More improvements/enhancements will be added in the upcoming FY.

## Impact
Provide an example of evaluating TRISO particle fuel performance based on Multiphysics simulation of a specific reactor design.

### PR DESCRIPTION
**Pull request by miaoyinb**
closes #327

### CODE REVIEW

**Comment by eshemon** [comment]
Thanks! @GiudGiud this is in support of an M2 due 12/15 so if you have time to get this in by then it would be great.

File: microreactors/mrad/triso_failure/TRISO/triso_sampler.i [hunk 1]
```diff
@@ -0,0 +1,4888 @@
...
+  [particle31_out]
+    type = CSV
+    file_base = results/particle31/
+    execute_on = 'TIMESTEP_END'
+  []
+  [particle32_out]
+    type = CSV
+    file_base = results/particle32/
+    execute_on = 'TIMESTEP_END'
+  []
```
Comment (GiudGiud) [inline microreactors/mrad/triso_failure/TRISO/triso_sampler.i:4887]: if you have dev funding for it this input could be 4888 / 32 = 153 lines 😆

File: doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
+# Multiphysics Models
+
+*Point of Contact: Yinbin Miao (ymiao.at.anl.gov)*
+
+*Primary Contributors: Nicholas Fassino, Yinbin Miao, Kun Mo, Nicolas Stauff*
+
+*Model link: [HPMR Model](https://github.com/idaholab/virtual_test_bed/tree/devel/microreactors/mrad)*
```
Comment (GiudGiud) [inline doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md:7]:
```suggestion
*Model link: [HPMR Model](https://github.com/idaholab/virtual_test_bed/tree/main/microreactors/mrad)*
```

File: doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+The HP-MR TRISO fuel particle model used in this exercise is a one-dimensional simplification of the model detailed in [!citep](Stauff2021). The model is separated into two component inputs: (1) a main particle model and (2) a master input that employs the MOOSE MultiApp functionality and the Samplers system to sample a user-specified number of simulations of the particle model. Particle layer widths are provided in the master input file, which converts them to normal distributions with user-provided standard deviations and bounds. Upon each execution of the particle simulation, these geometry distributions are sampled, and a set of layer coordinates is passed to the particle simulation with the Samplers system. Following the particle simulation, layer failure data are extracted from the particle simulation with the Transfers system. This information includes, for example, binary flags for multiple failure modes, the maximum neutron fluence in the particle, and fluence at failure (if applicable). The failure rate of the particle due to several mechanisms is then reported with VectorPostprocessors that track the stochastic results of the simulation.
```
Comment (GiudGiud) [inline doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md:14]:
```suggestion
The HP-MR TRISO fuel particle model used in this exercise is a one-dimensional simplification of the model detailed in [!citep](Stauff2021). The model is separated into two component inputs: (1) a main particle model and (2) a parent application input that employs the MOOSE MultiApp functionality and the Samplers system to sample a user-specified number of simulations of the particle model. Particle layer widths are provided in the parent input file, which converts them to normal distributions with user-provided standard deviations and bounds. Upon each execution of the particle simulation, these geometry distributions are sampled, and a set of layer coordinates is passed to the particle simulation with the Samplers system. Following the particle simulation, layer failure data are extracted from the particle simulation with the Transfers system. This information includes, for example, binary flags for multiple failure modes, the maximum neutron fluence in the particle, and fluence at failure (if applicable). The failure rate of the particle due to several mechanisms is then reported with VectorPostprocessors that track the stochastic results of the simulation.
```

File: doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+!listing /mrad/triso_failure/TRISO/triso_particle.i block=Functions/high_fidelity_strength_asphericity 
+
+!listing /mrad/triso_failure/TRISO/triso_particle.i block=Functions/stress_correlation_asphericity 
+
+!listing /mrad/triso_failure/TRISO/triso_particle.i block=Functions/stress_change_correlation_asphericity 
+
+## HP-MR TRISO MultiApps Model Setup
+
+MOOSE-based TRISO failure analysis in this demonstration is conducted with two BISON inputs in a MultiApps setup: (1) a TRISO sampler input serving as the master BISON app and (2) a TRISO particle input as the sub-app.
```
Comment (GiudGiud) [inline doc/content/microreactors/mrad/hpmr_triso_failure/problem_models.md:35]:
```suggestion
MOOSE-based TRISO failure analysis in this demonstration is conducted with two BISON inputs in a MultiApps setup: (1) a TRISO sampler input serving as the parent BISON app and (2) a TRISO particle input as the sub-app.
```

**Review by GiudGiud — COMMENTED** [review]
I ll make the changes and add the regression testing tomorrow

File: microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i [hunk 5]
```diff
@@ -0,0 +1,1693 @@
...
+                  top_reflector_1 top_reflector_2 top_reflector_3 top_reflector_4
+                  bottom_reflector_1 bottom_reflector_2 bottom_reflector_3 bottom_reflector_4'
+hp_surfs = 'HP_surf'
+
+[GlobalParams]
+  flux_conversion_factor = 1
+[]
+
+[Mesh]
+  file = ../mesh/3D_unit_cell_FY21_simple_YAN_40sections_5cm_axial_dist_bison.e
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i:72]: It is a mesh created a while ago...

File: microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i [hunk 6]
```diff
@@ -0,0 +1,1693 @@
...
+  petsc_options = '-snes_ksp_ew'
+  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -ksp_gmres_restart'
+  petsc_options_value = 'lu       superlu_dist                  51'
+  line_search = 'none'
+
+  nl_abs_tol = 1e-7
+  nl_rel_tol = 1e-7
+
+  nl_max_its = 20
+  fixed_point_max_its = 1
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i:624]: As it majority of the time was spending on the iterations for neutronics, this `1` here actually works well.

File: microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i [hunk 7]
```diff
@@ -0,0 +1,1693 @@
...
+    outputs = 'pp_fuelsampler_temp_32'
+    execute_on = 'initial timestep_end'
+  []
+  [axial_pow_32]
+    type = ElementAverageValue
+    variable = 'power_density'
+    block = fuel_32
+    outputs = 'pp_fuelsampler_pow_32'
+    execute_on = 'initial timestep_end'
+  []
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply microreactors/mrad/triso_failure/UnitCell_with_TM/MP_ss_bison.i:1172]: Need to transfer them as pp's

File: microreactors/mrad/triso_failure/UnitCell_with_TM/Exec_MP_ss_griffin.i [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+    multi_app = bison_diff
+    variable = power_density
+    source_variable = power_density
+    execute_on = 'initial timestep_end'
+    displaced_source_mesh = false
+    displaced_target_mesh = false
+    use_displaced_mesh = false
+  []
+  [from_sub_tempf]
+    type = MultiAppInterpolationTransfer
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply microreactors/mrad/triso_failure/UnitCell_with_TM/Exec_MP_ss_griffin.i]: I think it was copied from a previous input. Any difference between those two approaches?

File: microreactors/mrad/triso_failure/tests [hunk 9]
```diff
@@ -0,0 +1,47 @@
+[Tests]
+  # Syntax Check
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply microreactors/mrad/triso_failure/tests:2]: Will leave to you as you mentioned, thanks.

File: microreactors/mrad/triso_failure/tests [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    type = RunApp
+    input = 'UnitCell_with_TM/MP_ss_bison_TM.i'
+    check_input = True
+    executable_pattern = 'bison*|blue_crab*|dire_wolf*'
+  []
+  [fluids_sockeye_syntax]
+    type = RunApp
+    input = 'UnitCell_with_TM/MP_ss_sockeye.i'
+    check_input = True
+    executable_pattern = 'blue_crab*|sockeye*|dire_wolf*'
```
Comment (GiudGiud) [inline microreactors/mrad/triso_failure/tests:28]:
```suggestion
    executable_pattern = 'sockeye*|dire_wolf*'
```

**Comment by miaoyinb** [comment]
> Sockeye input seg faults locally. I ll try with a newer Sockeye then this might need some real debugging

It was working for the version on the day I submitted it...

**Comment by GiudGiud** [comment]
Works with current sockeye. I dont keep my machine that up to date.

**Comment by GiudGiud** [comment]
Actually no it's when you forget to use `--app SockeyeApp` with a BC executable.
This should be fixed soon with the `[Application]` syntax

File: microreactors/mrad/triso_failure/UnitCell_with_TM/Exec_MP_ss_griffin.i (hunk 8 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply microreactors/mrad/triso_failure/UnitCell_with_TM/Exec_MP_ss_griffin.i]: alright let s leave it like that for now. 
The difference is that you are using old code and it wrong grow

**Comment by GiudGiud** [comment]
Testing all good

**Review by GiudGiud — APPROVED** [review]
I think it would be good to spend time vectorizing the inputs a little.

Thank you for the contribution. It's a real asset for the VTB to host the progression of these models.

**Comment by miaoyinb** [comment]
Thank you so much, @GiudGiud.

**Pull request merged**